File: test/ctest/case/contractcall.js
'use strict';

function init(input_str) {
    Utils.assert(Chain.msg.asset === undefined);
}

//first,address-1,address-2,address-3
function main(input_str) {
    let a = input_str.split(',');
    if (a[0] === '1') {
        Chain.store('init-main', input_str);
        let str = input_str;
        let result = str.replace('1', '2');
        Chain.store('init-main-1', result);
        Chain.contractCall(a[1], true, '123', result);
    }

    if (a[0] === '2') {
        Chain.store('init-main', input_str);
        let str_2 = input_str;
        let result_2 = str_2.replace('2', '3');
        Chain.store('init-main-2', result_2);
        Chain.contractCall(a[2], true, '123', result_2);
    }

    if (a[0] === '3') {
        Chain.store('init-main', input_str);
        let str_3 = input_str;
        let result_3 = str_3.replace('3', '4');
        Chain.store('init-main-3', result_3);
        Chain.contractCall(a[3], true, '123', result_3);

        Chain.store('Chain.block.timestamp-main', Chain.block.timestamp.toString());
        Chain.store('Chain.block.number-main', Chain.block.number.toString());

        Chain.store('Chain.tx.initiator-main', Chain.tx.initiator);
        Chain.store('Chain.tx.sender-main', Chain.tx.sender);
        Chain.store('Chain.tx.gasPrice-main', Chain.tx.gasPrice.toString());
        Chain.store('Chain.tx.hash-main', Chain.tx.hash);
        Chain.store('Chain.tx.feeLimit-main', Chain.tx.feeLimit.toString());

        Chain.store('Chain.msg.initiator-main', Chain.msg.initiator);
        Chain.store('Chain.msg.sender-main', Chain.msg.sender);
        Chain.store('Chain.msg.coinAmount-main', Chain.msg.coinAmount.toString());
        Chain.store('Chain.msg.nonce-main', Chain.msg.nonce.toString());
        Chain.store('Chain.msg.operationIndex-main', Chain.msg.operationIndex.toString());
        Chain.store('Chain.thisAddress-main', Chain.thisAddress.toString());

        if (Chain.msg.asset !== undefined) {
            Chain.store('Chain.msg.asset-main', JSON.stringify(Chain.msg.asset));
        }
    }

    if (a[0] === '4') {
        if (a[4] !== undefined) {
            Chain.store('init-main-4', input_str);
        }
    }

    return '';
}

function query(input_str) {
    return '';
}

File: test/ctest/case/contractcall.cpp
#include "ctest.h"
#include "gtest/gtest.h"

class ContractCall : public testing::Test{
protected:
	void test();
	bumo::CTest ctest_;
};

TEST_F(ContractCall, test){ test(); }

void ContractCall::test(){
	std::string contract_address_0;
	std::string contract_address_1;
	std::string contract_address_2;
	std::string contract_address_3;

	bumo::CTest::Param param;
	param.balance = "10000000000";
	param.fee_limit = "20000000000";
	param.gas_price = "1000";
	param.input = "init";

	EXPECT_EQ(ctest_.Create(__FILE__, param), 0);
	contract_address_0 = ctest_.GetContractAddress();

	EXPECT_EQ(ctest_.Create(__FILE__, param), 0);
	contract_address_1 = ctest_.GetContractAddress();

	EXPECT_EQ(ctest_.Create(__FILE__, param), 0);
	contract_address_2 = ctest_.GetContractAddress();

	EXPECT_EQ(ctest_.Create(__FILE__, param), 0);
	contract_address_3 = ctest_.GetContractAddress();

	//dep 3
	param.input = "1," + contract_address_1 + "," + contract_address_2 + "," + contract_address_3 + ",error";
	EXPECT_EQ(ctest_.Paycoin(param, contract_address_0), 153);


	param.input = "1," + contract_address_1 + "," + contract_address_2 + "," + contract_address_3;
	EXPECT_EQ(ctest_.Paycoin(param, contract_address_0), 0);

	EXPECT_GT(utils::String::Stoi64(ctest_.Load("Chain.block.timestamp-main", contract_address_2)), 111111);
	EXPECT_GT(utils::String::Stoi64(ctest_.Load("Chain.block.number-main")), 1);
	EXPECT_EQ(ctest_.Load("Chain.tx.initiator-main"), ctest_.GetSender());
	EXPECT_EQ(ctest_.Load("Chain.tx.sender-main"), ctest_.GetSender());
	EXPECT_EQ(ctest_.Load("Chain.tx.gasPrice-main"), param.gas_price);
	EXPECT_EQ(ctest_.Load("Chain.tx.hash-main"), ctest_.GetLastHash());
	EXPECT_EQ(ctest_.Load("Chain.tx.feeLimit-main"), param.fee_limit);
	EXPECT_EQ(ctest_.Load("Chain.msg.initiator-main"), contract_address_1);
	EXPECT_EQ(ctest_.Load("Chain.msg.sender-main"), contract_address_1);
	EXPECT_EQ(ctest_.Load("Chain.msg.coinAmount-main"), "123");
	EXPECT_EQ(ctest_.Load("Chain.msg.nonce-main"), "3");
	EXPECT_EQ(ctest_.Load("Chain.msg.operationIndex-main"), "0");
	EXPECT_EQ(ctest_.Load("Chain.thisAddress-main"), ctest_.GetContractAddress());
	EXPECT_EQ(ctest_.Load("Chain.msg.asset-main"), "");
}
